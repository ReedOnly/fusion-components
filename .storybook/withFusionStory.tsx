import React from 'react';
import Header from '../src/components/core/Header';
import Content from '../src/components/core/Content';
import Overlay from '../src/components/general/ApplicationGuidance/components/Overlay';

export default (title: string, padding: number = 16) => (stories: () => React.ReactNode) => {
    return (
        <>
            <Overlay fixed>
                <Header aside={null} start={null} content={null} />
            </Overlay>
            <Content>
                <Overlay>
                    <div
                        style={{
                            padding,
                            minHeight: '100%',
                            width: '100%',
                            maxWidth: '100%',
                            boxSizing: 'border-box',
                        }}
                    >
                        {stories()}
                    </div>
                </Overlay>
            </Content>
        </>
    );
};
