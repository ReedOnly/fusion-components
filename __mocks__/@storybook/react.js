import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
