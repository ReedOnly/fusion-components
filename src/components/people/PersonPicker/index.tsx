import React, { useState, useEffect, useCallback } from 'react';
import {
    PersonPhoto,
    SearchableDropdownWrapper,
    SearchableDropdownSection,
    useTooltipRef,
} from '@equinor/fusion-components';

import { PersonDetails } from '@equinor/fusion';
import usePersonQuery from '../usePersonQuery';
import peopleToSections from './peopleToDropdownSections';

export type PersonPickerOption = {
    title: string;
    key: string;
    isSelected?: boolean;
    isDisabled?: boolean;
};

type PersonPickerProps = {
    onSelect?: (item: PersonPickerOption) => void;
};

const ItemComponent = ({ item }) => {
    const tooltipRef = useTooltipRef(item.person ? item.person.mail : '');

    if (item.person) {
        return <div ref={tooltipRef}>{item.title}</div>;
    }

    return item.title;
};

const AsideComponent = ({ item }) => {
    if (item.key === 'empty') {
        return null;
    }

    return <PersonPhoto person={item.person} size="medium" hideTooltip />;
};

export default ({ onSelect }: PersonPickerProps) => {
    const [sections, setSections] = useState<SearchableDropdownSection[]>([]);
    const [error, isQuerying, people, search] = usePersonQuery();
    const [searchQuery, setSearchQuery] = useState('');
    const [peopleMatch, setPeopleMatch] = useState<PersonDetails[]>([]);
    const [selectedPersonId, setSelectedPersonId] = useState('');

    useEffect(() => {
        search(searchQuery);
    }, [searchQuery]);

    useEffect(() => {
        setPeopleMatch(people.splice(0, 10));
    }, [people]);

    useEffect(() => {
        setSections(peopleToSections(peopleMatch, selectedPersonId, searchQuery, isQuerying));
    }, [peopleMatch, searchQuery, selectedPersonId, isQuerying]);

    const handleSelect = useCallback(item => {
        setSelectedPersonId(item.key);

        if (onSelect) {
            onSelect(item);
        }
    }, []);

    return (
        <SearchableDropdownWrapper
            sections={sections}
            onSelect={handleSelect}
            onSearchAsync={query => setSearchQuery(query)}
            itemComponent={ItemComponent}
            asideComponent={AsideComponent}
            label="Select person"
        />
    );
};
