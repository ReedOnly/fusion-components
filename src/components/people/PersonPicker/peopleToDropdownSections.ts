import { PersonDetails, PersonAccountType } from '@equinor/fusion';
import { SearchableDropdownSection } from '@equinor/fusion-components';

const filterPeople = (people: PersonDetails[], includedAccountTypes: PersonAccountType[]) => {
    return people.reduce((acc: PersonDetails[], curr: PersonDetails) => {
        const include = includedAccountTypes.indexOf(curr.accountType) !== -1;

        if (include) {
            acc.push(curr);
        }

        return acc;
    }, []);
};

export default function(
    people: PersonDetails[],
    selectedId: string,
    searchQuery: string,
    isQuerying: boolean
): SearchableDropdownSection[] {
    const primary = filterPeople(people, ['Consultant', 'Employee']);
    const secondary = filterPeople(people, ['External']);

    const isNoMatches = primary.length === 0 && secondary.length === 0;

    if ((isNoMatches && searchQuery !== '') || isQuerying) {
        return [
            {
                key: 'nomatches',
                items: [
                    {
                        key: 'empty',
                        title: isQuerying ? 'Searching...' : 'No results.',
                        isDisabled: true,
                    },
                ],
            },
        ];
    }

    const sections: SearchableDropdownSection[] = [];

    if (primary.length > 0) {
        sections.push({
            key: 'primary',
            title: 'Employees and consultants',
            items: primary.map((p: PersonDetails) => ({
                key: p.azureUniqueId,
                title: p.name,
                person: p,
                isSelected: p.azureUniqueId === selectedId,
            })),
        });
    }

    if (secondary.length > 0) {
        sections.push({
            key: 'secondary',
            title: 'External',
            items: secondary.map((p: PersonDetails) => ({
                key: p.azureUniqueId,
                title: p.name,
                person: p,
                isSelected: p.azureUniqueId === selectedId,
            })),
        });
    }

    return sections;
}
