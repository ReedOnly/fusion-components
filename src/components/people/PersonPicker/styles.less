@import '~styles/mixins.less';

.container {
    max-width: .grid-unit(41px)[];
}

.selectedPerson {
    display: flex;
    justify-content: space-between;
    min-height: .grid-unit(5px)[];
    max-height: .grid-unit(7px)[];
    max-width: .grid-unit(41px)[];
    background-color: var(--color-black-alt5);
    border-bottom: 1px solid var(--color-primary-accent);
    caret-color: var(--color-primary);
    padding: .grid-unit(1px)[] .grid-unit(2px)[];
    border-radius: 4px 4px 0 0;
    align-items: center;

    .content {
        display:flex;
        flex-direction: row;
        align-items: center;

        .name {
            padding-left: .grid-unit(1px)[];
        }
    }

    .icon {
        align-self: center;
        padding-left: .grid-unit(2px)[];

        svg {
            width: .grid-unit(3px)[];
            height: .grid-unit(3px)[];
        }
    }
}