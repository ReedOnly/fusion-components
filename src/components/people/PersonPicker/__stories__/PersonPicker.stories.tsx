import React from 'react';
import { storiesOf } from '@storybook/react';
import withFusionStory from '../../../../../.storybook/withFusionStory';
import PersonPicker from '../index';

storiesOf('People|PersonPicker', module)
    .addParameters({ jest: ['PersonPicker.stories.jsx'] })
    .addDecorator(withFusionStory('PersonPicker'))
    .add('Default', () => <PersonPicker />);
