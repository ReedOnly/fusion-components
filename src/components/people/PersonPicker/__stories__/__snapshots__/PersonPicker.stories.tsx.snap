// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`People|PersonPicker Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div>
      <div
        className="inputContainer"
      >
        <div
          className="inputContent labelLess"
          onClick={[Function]}
        >
          <div
            className="inputTextContent"
          >
            <input
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyUp={[Function]}
              placeholder="Find person"
              value=""
            />
          </div>
          <div
            className="icon"
            onClick={[Function]}
          >
            <svg
              cursor="pointer"
              height={24}
              style={
                Object {
                  "color": "",
                }
              }
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M8.70627 11.4137L11.2963 14.0037C11.6863 14.3937 12.3163 14.3937 12.7063 14.0037L15.2963 11.4137C15.9263 10.7837 15.4763 9.70374 14.5863 9.70374H9.40627C8.51627 9.70374 8.07627 10.7837 8.70627 11.4137Z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`People|PersonPicker Pre Selected 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div>
      <div
        className="inputContainer"
      >
        <div
          className="inputContent"
          onClick={[Function]}
        >
          <div
            className="inputTextContent"
          >
            <label>
              Selected person
            </label>
            <input
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyUp={[Function]}
              placeholder=""
              value=""
            />
          </div>
          <div
            className="icon"
            onClick={[Function]}
          >
            <svg
              cursor="pointer"
              height={24}
              style={
                Object {
                  "color": "",
                }
              }
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M8.70627 11.4137L11.2963 14.0037C11.6863 14.3937 12.3163 14.3937 12.7063 14.0037L15.2963 11.4137C15.9263 10.7837 15.4763 9.70374 14.5863 9.70374H9.40627C8.51627 9.70374 8.07627 10.7837 8.70627 11.4137Z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
